add new feature: dump dependent Tasks information as a table

mski-iksm opened the pull request:
I've added a new feature for dumping task cache information file at `gokart.build()`.

Task cache information file is a pandas dataframe which has information such as task name, unique id, cache file path, parameter and processing time of completed tasks.

issue: https://github.com/m3dev/gokart/issues/231

mski-iksm commented:
@Hi-king @vaaaaanquish @e-mon @hirosassa Please review.

hirosassa commented on docs/task_information.rst:
```diff
@@ -170,3 +170,19 @@ the output could be like:
+
+Dump TaskInfo Table
+--------------------
+When using ``gokart.build()``, you can dump information of dependent tasks to a file as TaskInfo table.
```
[ask] Is it difficult to dump the information on standard out or error as a log, too?

mski-iksm replied:
@hirosassa 
If you want to dump the information as a standard out, I think of using `make_tree_info` may be better.

vaaaaanquish commented:
Since these information can be recovered from log that are already there, looks unnecessary feature.
Please write usage scenarios in document.

vaaaaanquish commented on gokart/task_info.py:
```diff
@@ -0,0 +1,57 @@
+
+def dump_task_info_table(task: TaskOnKart, task_info_dump_path: Optional[str], task_info_ignore_task_names: Optional[List[str]]):
+    if task_info_dump_path is None:
+        return None
```
It's better to make the decision outside of function.

vaaaaanquish commented on docs/task_information.rst:
```diff
@@ -170,3 +170,19 @@ the output could be like:
+
+.. code:: python
+
+    gokart.build(task=task, task_info_dump_path="./task_info.csv", task_info_ignore_task_names=None)
```
Write a description of `task_info_ignore_task_names` to here.

vaaaaanquish commented on docs/task_information.rst:
```diff
@@ -170,3 +170,19 @@ the output could be like:
+To dump TaskInfo table, set ``task_info_dump_path`` parameter at ``gokart.build()``.
+
+- path destination can be `local`, `S3`, or `GCS`
+- file extension can be any type that gokart file processor accepts, including `csv`, `pickle`, or `txt`.
```
Please add a link to the file proceccer documentation.

vaaaaanquish replied:
for example: https://gokart.readthedocs.io/en/latest/task_on_kart.html#taskonkart-make-target

mski-iksm commented:
@vaaaaanquish 

> Since these information can be recovered from log that are already there, looks unnecessary feature.

I think you are talking about using `gokart.make_tree_info()` which is very close feature.
However,
- it's not formatted and difficult to read
- In some case, getting the root task with all parameters set is difficult to reproduce

so, I think this feature is needed for productivity.


I do agree with `_make_task_info_list()` duplicated with  `gokart.make_tree_info()`, so I think replacing backend with `gokart.make_tree_info()`. How do you think?

vaaaaanquish commented:
@mski-iksm 
> I do agree with _make_task_info_list() duplicated with gokart.make_tree_info(), so I think replacing backend with gokart.make_tree_info()

This is very cool. If `make_tree_info` has the ability to return `dataframes`, It's a big implement.

mski-iksm commented:
@vaaaaanquish 
I've modified `dump_task_info_table` to use the same backend as `make_tree_info`.
`dump_task_info_table` is mostly same with `make_tree_info`, except it dumps the tree_info as pandas.DataFrame.

During this process, I've refactored `make_tree_info` and fixed some tests.

mski-iksm commented:
I'll close this PR and separate it into following 2 PRs.
- Refactor `make_tree_info` + move to different package
- implement feature of dumping task info table as a option of `make_tree_info`
